### PR TITLE
Kafka-clients: records(TopicPartition): unit tests + some (potential) improvements

### PR DESCRIPTION
**Pull request by gygabyte**
I think these instrumented methods should safeguard for misuse and therefore prevent inaccurate traces. Since these methods can only be instrumented when applications calls them as it is imposed by the synchronous nature of the Kafka-client, then we should ensure (up to certain point) the agent collects what exactly the Kafka-client has processed. One example is,  for instance, the application calls the iterator multiple times.
subList should not be instrumented as the application can remove elements of the list (as the method implies) and therefore, again, wrong traces can be collected which won't be the real representation of what has been received by the client

### CODE REVIEW

File: dd-java-agent/instrumentation/kafka-clients-0.11/src/test/groovy/KafkaClientTest.groovy [hunk 1]
```diff
@@ -119,6 +124,94 @@ class KafkaClientTest extends AgentTestRunner {
...
+    def producer = new KafkaProducer(senderProps)
+
+    consumer.assign(Arrays.asList(new TopicPartition(SHARED_TOPIC, kafkaPartition)))
+
+    when:
+    def greeting = "Hello from MockConsumer!"
+    producer.send(new ProducerRecord<Integer, String>(SHARED_TOPIC, kafkaPartition, null, greeting))
+    TEST_WRITER.waitForTraces(1)
+
+    then:
```
Comment (mar-kolya) [inline dd-java-agent/instrumentation/kafka-clients-0.11/src/test/groovy/KafkaClientTest.groovy]: Looks like you may be able to simplify this to something like:

```
    when:
    def greeting = "Hello from MockConsumer!"
    producer.send(new ProducerRecord<Integer, String>(SHARED_TOPIC, kafkaPartition, null, greeting))

    then:
    TEST_WRITER.waitForTraces(1) // ensure consistent ordering of traces
    def pollResult = KafkaTestUtils.getRecords(consumer)
    def records = pollResult.records(new TopicPartition(SHARED_TOPIC, kafkaPartition)).iterator()

    def first
    while (records.hasNext()) {
      first = records.next()
      break
    }
    records.hasNext() == false
    first.value() == greeting
    first.key() == null
```

File: dd-java-agent/instrumentation/kafka-clients-0.11/src/main/java/datadog/trace/instrumentation/kafka_clients/TracingList.java [hunk 2]
```diff
@@ -121,6 +113,11 @@ public int lastIndexOf(final Object o) {
     return delegate.lastIndexOf(o);
   }
 
+  @Override
+  public Iterator<ConsumerRecord> iterator() {
+    return super.iterator();
+  }
```
[the remark this answers is not in the document]
Comment (gygabyte) [reply dd-java-agent/instrumentation/kafka-clients-0.11/src/main/java/datadog/trace/instrumentation/kafka_clients/TracingList.java]: Yep, I understand, but I was getting compiler warnings regarding @NotNull annotation.. not a big deal ?

[the remark this answers is not in the document]
Comment (mar-kolya) [reply dd-java-agent/instrumentation/kafka-clients-0.11/src/main/java/datadog/trace/instrumentation/kafka_clients/TracingList.java]: No. you can leave this as is if you prefer

[the remark this answers is not in the document]
Comment (gygabyte) [reply dd-java-agent/instrumentation/kafka-clients-0.11/src/main/java/datadog/trace/instrumentation/kafka_clients/TracingList.java]: I will take it out, we can avoid a call in the stack. thanks!

File: LICENSE-3rdparty.csv [hunk 3]
```diff
@@ -9,3 +9,4 @@ logback.xml,net.logstash.logback,Apache-2.0,
 import (test),org.junit,EPL-1.0,Copyright © 2002-2017 JUnit. All Rights Reserved.
 import (test),org.assertj,Apache-2.0,Copyright 2012-2017 the original author or authors.
 import (test),org.mockito,MIT,Copyright (c) 2007 Mockito contributors
+kafka-clients: method records(TopicPartition) partial+test,"Aspect Software, Inc.", Apache-2.0, "Copyright (C) Aspect Software, Inc."
```
Comment (mar-kolya) [inline LICENSE-3rdparty.csv]: I do not think this change is needed. I think this file is specifically for external deps that we ship with the built agent. This change doesn't add new dependencies.

Comment (gygabyte) [reply LICENSE-3rdparty.csv]: sure, this is more about the contribution as I/we thought it was what this is file is for. Is there anyway of the contribution being recognised in some other place? Thanks!

File: dd-java-agent/instrumentation/kafka-clients-0.11/src/test/groovy/KafkaClientTest.groovy (hunk 1 above)
Comment (gygabyte) [reply dd-java-agent/instrumentation/kafka-clients-0.11/src/test/groovy/KafkaClientTest.groovy]: I will change the WAIT_TRACES of producer to the `then` block:

On other note,  I don't need to wait for traces when consuming?

Comment (mar-kolya) [reply dd-java-agent/instrumentation/kafka-clients-0.11/src/test/groovy/KafkaClientTest.groovy]: `assertTraces(2) {` waits to see exactly two traces. If I understand things correctly you need `TEST_WRITER.waitForTraces(1)` only once right before consuming - to make sure producer traces has been shipped - so two traces do not appear out of order at assertion time.

Comment (gygabyte) [reply dd-java-agent/instrumentation/kafka-clients-0.11/src/test/groovy/KafkaClientTest.groovy]: yep makes sense. I have updated the PR as per your comments

File: LICENSE-3rdparty.csv (hunk 3 above)
Comment (tylerbenson) [reply LICENSE-3rdparty.csv]: We will certainly give you a shout-out in our release notes.  Thanks again for taking the time and effort to contribute.

Comment (gygabyte) [reply LICENSE-3rdparty.csv]: Appreciate it! Happy to contribute!

**Pull request merged**